lib/semantics/resolve-names.cc

```diff
@@ -1037,11 +1037,271 @@ class ConstructVisitor : public DeclarationVisitor {
+  }
+  bool find_dataSharingAttributeObject(const Symbol *);
+
+protected:
+  // TODO: resolve variables referenced in the OpenMP region
+  void ResolveOmpObjectList(const parser::OmpObjectList &, Symbol::Flag);
+  void ResolveOmpObject(const parser::OmpObject &, Symbol::Flag);
+  Symbol *ResolveOmp(const parser::Name &, Symbol::Flag);
+  Symbol *ResolveOmp(Symbol &, Symbol::Flag);
+  Symbol *ResolveOmpCommonBlockName(const parser::Name *);
+  Symbol *DeclarePrivateAccessEntity(const parser::Name &, Symbol::Flag);
+  Symbol *DeclarePrivateAccessEntity(Symbol &, Symbol::Flag);
+  Symbol *DeclareOrMarkOtherAccessEntity(const parser::Name &, Symbol::Flag);
+  Symbol *DeclareOrMarkOtherAccessEntity(Symbol &, Symbol::Flag);
+  void CheckMultipleAppearances(
+      const parser::Name &, const Symbol *, Symbol::Flag);
+
+private:
+  std::set<const Symbol *> dataSharingAttributeObjects_;  // on one directive
+};
+
+bool OmpVisitor::find_dataSharingAttributeObject(const Symbol *object) {
+  auto it{dataSharingAttributeObjects_.find(object)};
+  if (it != dataSharingAttributeObjects_.end()) {
+    return true;
+  }
+  return false;
+}
+
+Symbol *OmpVisitor::ResolveOmpCommonBlockName(const parser::Name *name) {
+  if (auto *prev{name ? currScope().parent().FindCommonBlock(name->source)
+                      : nullptr}) {
+    name->symbol = prev;
+    return prev;
+  } else {
+    return nullptr;
+  }
+}
+
+void OmpVisitor::ResolveOmpObjectList(
+    const parser::OmpObjectList &ompObjectList, Symbol::Flag ompFlag) {
+  for (const auto &ompObject : ompObjectList.v) {
+    ResolveOmpObject(ompObject, ompFlag);
+  }
+}
+
+void OmpVisitor::ResolveOmpObject(
+    const parser::OmpObject &ompObject, Symbol::Flag ompFlag) {
+  if (const auto *designator{std::get_if<parser::Designator>(&ompObject.u)}) {
+    if (const auto *name{GetDesignatorNameIf(*designator)}) {
+      auto *symbol{ResolveOmp(*name, ompFlag)};
+      if (dataSharingAttributeFlags.test(ompFlag)) {
+        CheckMultipleAppearances(*name, symbol, ompFlag);
+      }
+    } else if (const auto *designatorName{ResolveDesignator(*designator)};
+               designatorName->symbol) {
+      // Array sections to be changed to substrings as needed
+      if (AnalyzeExpr(context(), *designator)) {
+        if (const auto *substring{
+                std::get_if<parser::Substring>(&designator->u)}) {
+          Say(designator->source,
+              "Fortran Substrings are not allowed on OpenMP "
+              "directives or clauses"_err_en_US);
+        }
+      }
+      // other checks, more TBD
+      if (const auto *details{
+              designatorName->symbol->detailsIf<ObjectEntityDetails>()}) {
+        if (details->IsArray()) {
+          // TODO: check Array Sections
+        } else if (designatorName->symbol->owner().IsDerivedType()) {
+          // TODO: check Structure Component
+        }
+      }
+    }
+  } else if (const auto *name{std::get_if<parser::Name>(&ompObject.u)}) {
```

Comment:
Doesn't it have to be a `parser::Name` at this point?

If you use `std::visit` instead of `if (... std::get_if` you would get a compilation error if there are any cases you are missing.

Reply:
In this case actually it has to be `parser::Name`.  Now `OmpObject` defines as
```
struct OmpObject {
  UNION_CLASS_BOILERPLATE(OmpObject);
  std::variant<Designator, /*common block*/ Name> u;
};
```

Reply:
Right. You already checked it's not `Designator`, so doesn't it have to be `Name` now?

`std::visit` is still better.

Reply:
oh right right, I will change that. Thanks.

Reply:
I've replaced the `if...else...` with `std::visit`.